### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by Kaweesi-Matia**
## Below is the description of the PR

- [ ] Make sure that the N+1 problem is solved when fetching all posts and their comments for a user
- [ ] Use Capybara to write integration tests for each view in your project.
- [ ] User index page:
- I can see the username of all other users.
- I can see the profile picture for each user.
- I can see the number of posts each user has written.
- When I click on a user, I am redirected to that user's show page.
- 

- [ ] user show page:
- I can see the user's profile picture.

- I can see the user's username.

- I can see the number of posts the user has written.
- I can see the user's bio.
- I can see the user's first 3 posts.
- 
blog-1
blog-2
blog-3
blog-4
I can see a button that lets me view all of a user's posts.
When I click a user's post, it redirects me to that post's show page.
When I click to see all posts, it redirects me to the user's post's index page
before
after

### CODE REVIEW

**Comment by addod19** [comment]
### Hi @Kaweesi-Matia, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a project, that is invalid because the linters are failing

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: spec/integration/user_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+      expect(page).to have_text(user.Name)
+      expect(page).to have_content('Number of posts')
+
+      expect(page).to have_text(user.PostsCounter)
+    end
+  end
+  it 'shows the number  of  posts for the user' do
+    visit '/users'
+    expect(page).to have_content('Number of posts')
+  end
```
Comment (t-yanick) [inline spec/integration/user_index_spec.rb:40]: - [ ] Great job adding some of the tests for user index page as per project requirements. You are missing the the `expect` statement to check if a user has a photo. Kindly add it.
- [ ] You are supposed to add a test that redirects to the user's show page when you click on a user. This test is missing for your user index page. Please kindly add it.

File: spec/integration/user_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  end
+  it 'displays page header' do
+    visit user_path(User.last)
+    expect(page).to have_text('User has no posts yet')
+  end
+
+  it 'displays page button' do
+    visit user_path(User.last)
+    expect(page).to have_text('See all posts')
+  end
```
Comment (t-yanick) [inline spec/integration/user_show_spec.rb:55]: - [ ] As per project requirements, you are expected to add two redirects tests:
- Test that redirects to the post's show page when you click on a user's post.
- Test that redirects to the user's post's index page when you click on a **see all posts**.
Kindly add these two tests.

File: spec/integration/post_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    expect(page).to have_content(@first_post.Text)
+    expect(page).to have_content(@second_post.Text)
+    expect(page).not_to have_content('michael')
+  end
+
+  it 'shows post\'s first comment' do
+    visit "/users/#{@user1.id}/posts"
+    expect(page).to have_content(@first_post.Text)
+  end
+
```
Comment (t-yanick) [inline spec/integration/post_index_spec.rb:40]: - [ ] Great job adding these tests for post index page, but your are missing one important test specified in the project requirements:
- Test that asserts you can see how many comments a post has.
Please kindly add this test case.

File: spec/integration/post_index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    @user1 = User.create(Name: 'Rodrigo De Paul',
+                         photo: 'https://upload.wikimedia.org/wikipedia/commons/thumb/0/09/Andrew_Carnegie%2C_by_Theodore_Marceau.jpg/800px-Andrew_Carnegie%2C_by_Theodore_Marceau.jpg',
+                         Bio: 'Industrialist.',
+                         PostsCounter: 2)
+    @first_post = Post.create(Title: 'Helloween', Text: 'This is my first post', user_id: @user1.id,
+                              comments_counter: 3, likes_counter: 2)
+    @second_post = Post.create(Title: 'Musically', Text: 'Micheal Jackson is the best ever', user_id: @user1.id,
+                               comments_counter: 3, likes_counter: 2)
+  end
+
```
Comment (t-yanick) [inline spec/integration/post_index_spec.rb:13]: - [ ] You are missing the very first test case for post index page as per project requirements:
- Test that asserts you can see the user's profile picture.
Please do well to add this test case.

**Review by t-yanick — CHANGES_REQUESTED** [review]
Hi @Kaweesi-Matia , 

### Good job so far!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Good job making the linters to pass
:heavy_check_mark: Great use of capybara for your integration tests

There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️

- [ ] As per project requirements, you are expected `add screenshots of the logs before and after fix of the N + 1 problem to the PR description.` But the logs are absent in your PR. Remeber the logs are screenshots from the terminal.
- [ ] To solve the N + 1 problem when fetching all posts and their comments for a user, for example, posts#index actions should use `@posts = @user.posts.includes(:comments)` . instead of `@posts = @user.posts`. Kindly check your PostsController index action to ensure that you are fetching all posts and their comments correctly.
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- Capitalizing the first letters of your comit messages. There are already descriptive and in the imperative tense

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/integration/user_index_spec.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    @users = User.all
+  end
+  it 'shows all the properties (Name, photo, number of posts) of each user' do
+    visit users_path
+    @users.each do |user|
+      expect(page).to have_text(user.Name)
+      expect(page).to have_content('Number of posts')
+
+      expect(page).to have_text(user.PostsCounter)
+    end
```
Comment (HAMMAS-SALEEM) [inline spec/integration/user_index_spec.rb]: - [ ] As the previous reviewer suggested that you should have a test to check if the user has a photo or not :+1: Either you can add an expect in this block or you can make a new it block to check photo of a user :+1:

File: spec/integration/user_index_spec.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  end
```
Comment (HAMMAS-SALEEM) [inline spec/integration/user_index_spec.rb:30]: - [ ] As the previous reviewer suggested that you are missing an test and test is if you click on any of the user then it takes you to the user show page :+1: Kindly add a test for that :blush:

File: spec/integration/user_show_spec.rb (hunk 2 above)
Comment (HAMMAS-SALEEM) [inline spec/integration/user_show_spec.rb:55]: - [ ] As the previous reviewer suggested that you are missing these two tests :fearful:

* Test that redirects to the post's show page when you click on a user's post.
* Test that redirects to the user's post's index page when you click on see all posts.
**Kindly add these two tests.**

**Hint:**  You are supposed to user **click_link** or **click_button** to make these tests :+1:

File: spec/integration/user_show_spec.rb [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+end
```
Comment (HAMMAS-SALEEM) [inline spec/integration/user_show_spec.rb:56]: - [ ] As the previous reviewer suggested that your are missing two important tests specified in the project requirements:
* Test that asserts you can see how many comments a post has.
* Test that asserts you can see the user's profile picture.
**Please kindly add these test case. :blush:**

**Review by HAMMAS-SALEEM — CHANGES_REQUESTED** [review]
# Required Changes :recycle:
almost-there

## Hey @Kaweesi-Matia :wave:

## Highlights :1st_place_medal:

✔️ Descriptive PR Title
✔️ Descriptive PR Summary
✔️ Following Gitflow
✔️ Following some requirements

Good job on implementing some of the requested changes, there are some issues that you still need to work on to go to the next project but you are almost there! :blush:

# Optional

_All the changes with optional tags are not crucial enough to prevent you from getting the approval but we highly recommend to students that they should take these changes into consideration._

Cheers and Happy coding! :clap: :clap: :clap:

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

---------------------------------------

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/integration/post_index_spec.rb (hunk 4 above)
Comment (Kaweesi-Matia) [reply spec/integration/post_index_spec.rb:13]: resolve problem

Comment (Kaweesi-Matia) [reply spec/integration/post_index_spec.rb:13]: resolve problem

File: spec/integration/user_index_spec.rb (hunk 5 above)
Comment (Kaweesi-Matia) [reply spec/integration/user_index_spec.rb]: resolve problem

File: spec/integration/user_show_spec.rb (hunk 7 above)
Comment (Kaweesi-Matia) [reply spec/integration/user_show_spec.rb:56]: resolve problem

**Review by elisha2kyakpo1 — APPROVED** [review]
Hi 👋 @Kaweesi-Matia, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

✔️ The PR title and description look good and professional.
✔️ No linter errors.
✔️ Tests passing👏

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**